inline/src/parser/mod.rs

```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
+                            if !tkn.is_plain_whitespace() {
+                                self.token_cache = Some(tkn);
+                                break;
+                            }
+                        }
+                    }
+                } else {
+                    // cache popped token and break
+                    self.token_cache = Some(next_token);
+                    break;
+                }
+            } else {
+                let (next_content, next_span) = next_token.parts();
+                content.push_str(next_content);
+                span.end = next_span.end;
+            }
         }
 
-        while let Some(next_token) = self.next_token() {
-            end = next_token.span.end;
+        Inline::plain_or_eol(content, span, kind)
+    }
 
-            if next_token.closes() {
+    fn parse_nested(&mut self, start_token: Token) -> Inline {
+        let start = start_token.span.start;
+        let mut end = start_token.span.end;
+        let mut content = VecDeque::new();
+
+        while let Some(next_token) = self.next_token() {
+            if next_token.closes(Some(&start_token)) {
+                end = next_token.span.end;
                 break;
             } else if next_token.opens() {
                 // lexer resolved tokens, if token opens, it is guaranteed that closing exists too.
                 // If not, it's bug in implementation
                 let nested = self.parse_inline(next_token);
-                content.append_inline(nested);
+                content.push_back(nested);
             } else {
-                if kind == TokenKind::Plain && next_token.kind != TokenKind::Plain {
-                    self.token_cache = Some(next_token);
-                    break;
-                }
+                // opening token not yet closed, next token does not start a new inline -> parse as
+                // plain text
 
                 end = next_token.span.end;
-                content.append(InlineContent::from(next_token));
+                let (inner, span) = next_token.parts();
+                let inline = Inline::Plain(Plain {
+                    content: String::from(inner),
+                    span,
+                });
+                content.push_back(inline);
             }
         }
 
         let span = Span::from((start, end));
-        content.try_flatten();
-        Inline::with_span(content, kind, span)
+        Inline::nested_with_span(content, start_token.kind, span)
+    }
+
+    fn parse_inline(&mut self, token: Token) -> Inline {
+        // opening/closing of tokens is resolved at lexing stage
+        // at this point we can simply parse
+        let kind = token.kind;
+
+        let mut inline = match (kind, token.opens()) {
+            (TokenKind::Verbatim, _) => self.parse_plain(
+                token,
+                PlainContext {
+                    enclosed: true,
+                    merge_whitespace: false,
```

Comment:
I am not entirely sure if we should use this for some other inlines apart from verbatim, but otherwise you're right. Right now, that naming would make more sense.

Reply:
I am also unsure. Maybe adding doc-comments is sufficient enough.

Reply:
I will add doc-comments here and leave it as is for now.